cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -188,6 +201,81 @@ public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile
+   * @throws InvalidArtifactException if the artifact is invalid
+   * @throws NotFoundException Not Found exception thrown from plugin method invocation or if no matching plugin found.
+   * @throws InvocationTargetException exception thrown from plugin method invocation.
+   */
+  public String invokePluginEndpoint(PluginInstantiator pluginInstantiator, Id.Artifact artifact,
+                                     PluginClass pluginClass, String methodName, String methodParameter,
+                                     DefaultEndpointPluginContext endpointPluginContext)
+    throws IOException, BadRequestException, NotFoundException, ClassNotFoundException, InvocationTargetException {
+
+    // See if there are export packages. Plugins should be in those packages
+
+    ClassLoader pluginClassLoader = pluginInstantiator.getArtifactClassLoader(artifact.toArtifactId());
+
+    Class pluginClassLoaded = pluginClassLoader.loadClass(pluginClass.getClassName());
+
+    Method[] methods = pluginClassLoaded.getDeclaredMethods();
+    for (Method method : methods) {
+      Path pathAnnotation = method.getAnnotation(Path.class);
+      // method should have path annotation else continue
+      if (pathAnnotation == null || !pathAnnotation.value().equals(methodName)) {
+        continue;
+      }
+
+      // make configField name as null, as we might want to call these methods with incomplete config.
+      PluginClass pluginClassWithoutConfig = new PluginClass(pluginClass.getType(), pluginClass.getName(),
+                                                             pluginClass.getDescription(), pluginClass.getClassName(),
+                                                             null, pluginClass.getProperties(),
+                                                             pluginClass.getEndpoints());
+      co.cask.cdap.api.plugin.Plugin plugin =
+        new co.cask.cdap.api.plugin.Plugin(artifact.toArtifactId(), pluginClassWithoutConfig,
+                                           PluginProperties.builder().build());
+      try {
+        Object result;
+        if (method.getParameterTypes().length == 2 &&
+          method.getParameterTypes()[1].equals(EndpointPluginContext.class)) {
+          result = method.invoke(pluginInstantiator.newInstance(plugin),
+                                 GSON.fromJson(methodParameter, method.getParameterTypes()[0]),
+                                 endpointPluginContext);
+        } else {
+          result = method.invoke(pluginInstantiator.newInstance(plugin),
+                                 GSON.fromJson(methodParameter, method.getParameterTypes()[0]));
+        }
+
+        // serialize the result and return
+        return GSON.toJson(result, method.getReturnType());
+      } catch (InvocationTargetException e) {
+        LOG.error("Got Exception when invoking plugin method {}", e);
+        if (e.getCause() instanceof NotFoundException) {
```

Comment:
this error handling should be done by callers of this method, not inside this method.  You should just throw the InvocationTargetException, which is what the method is declaring anyway.  With this catch, you're actually never throwing an InvocationTargetException right?